src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs

```diff
@@ -146,15 +152,48 @@ public async Task<UpsertOutcome> UpsertAsync(ResourceWrapper resource, WeakETag
                     }
                 }
             }
+            finally
+            {
+                CloseConnectionIfNeeded(connection);
+            }
+        }
+
+        private SqlCommand CreateCommand(SqlConnection connection)
+        {
+            SqlCommand command = connection.CreateCommand();
+            command.Transaction = _sqlTransactionHandler.SqlTransactionScope?.SqlTransaction;
+            return command;
+        }
+
+        private void CloseConnectionIfNeeded(SqlConnection connection)
+        {
+            if (_sqlTransactionHandler.SqlTransactionScope?.SqlConnection == null)
+            {
+                connection.Close();
+            }
+        }
+
+        private SqlConnection ObtainSqlConnection()
```

Comment:
This would also be needed in search, if there are searches in the bundles, right?